### PR TITLE
feature/omniauth- FB, Google登入及帳號整合

### PR DESCRIPTION
**Pull request by haopingfan**
How to test:

在 localhotst 中實際執行Fb與google登入, 結果皆如同 callbacks_controller 之呈現。

### CODE REVIEW

File: app/controllers/callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  def callback_handler
+    auth = request.env['omniauth.auth']
+    @identity = Identity.find_or_intialize_with_omniauth(auth)
+
+    if @identity.user
+      sign_in @identity.user
+      redirect_to root_path, notice: 'Signed in. Welcome back!'
+    elsif User.find_by(email: auth.info['email'])
+      @user = User.find_by(email: auth.info['email'])
```
Comment (khiav223577) [inline app/controllers/callbacks_controller.rb]: 上一行已經 query 過同樣的東西了，這裡多花一個 query

File: app/controllers/callbacks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      set_identity_and_sign_in(@identity, @user, 'Successfully linked this account. Signed in. Welcome back!')
+    else
+      @user = User.initialize_with_omniauth(auth)
```
Comment (khiav223577) [inline app/controllers/callbacks_controller.rb]: 我會把這個函式改成 create

File: db/migrate/20170918073349_create_identities.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+class CreateIdentities < ActiveRecord::Migration[5.1]
+  def change
+    create_table :identities do |t|
+      t.string :provider
+      t.string :uid
```
Comment (khiav223577) [inline db/migrate/20170918073349_create_identities.rb]: 可以加些 index

File: db/migrate/20170918073349_create_identities.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class CreateIdentities < ActiveRecord::Migration[5.1]
+  def change
+    create_table :identities do |t|
+      t.string :provider, null: false, index: true
+      t.string :uid, null: false, index: true
```
Comment (khiav223577) [inline db/migrate/20170918073349_create_identities.rb]: `[:provider, :uid]` 這合起來應該要是 unique，可以加個  unique index

Comment (haopingfan) [reply db/migrate/20170918073349_create_identities.rb]: 更正囉 謝謝神明

File: config/initializers/devise.rb [hunk 5]
```diff
@@ -274,4 +274,6 @@
   # When using OmniAuth, Devise cannot automatically set OmniAuth path,
   # so you need to do it manually. For the users scope, it would be:
   # config.omniauth_path_prefix = '/my_engine/users/auth'
+  config.omniauth :facebook, '275091769672408', '336a1f69ca20f389830606193c00948c'
```
Comment (khiav223577) [inline config/initializers/devise.rb]: facebook api 的 secret 不要 commit 上來

Comment (haopingfan) [reply config/initializers/devise.rb]: 原來如此, 又學到一招了 耶! 感謝神明~~~ 已更新~

**Comment by khiav223577** [comment]
然後建議把 commit 紀錄砍掉，否則 git 仍然能夠搜尋到你的密鑰

**Comment by haopingfan** [comment]
已將commit紀錄更正, 謝謝神明！

File: config/secrets.yml [hunk 6]
```diff
@@ -1,32 +0,0 @@
...
-
-# Shared secrets are available across all environments.
-
-# shared:
-#   api_key: a1B2c3D4e5F6
-
-# Environmental secrets are only available for that specific environment.
-
-development:
-  secret_key_base: 1ea818af517ba312d965c487fc8ae2224dbe15ee88f19b4d3a15cb26557c1c5a9b01610bf57e516a2fa24d59b2234d5907efb1261beef035f27c72fee29b7c48
```
Comment (khiav223577) [inline config/secrets.yml]: 這沒有改掉XD，這也是機密資訊

**Comment by khiav223577** [comment]
還是被 github cache 住了XD
image

File: config/secrets.yml (hunk 6 above)
Comment (haopingfan) [reply config/secrets.yml]: 這也是機密資訊喔XD

Comment (haopingfan) [reply config/secrets.yml]: 好喔～

**Comment by haopingfan** [comment]
對啊 應該是留言的關係 這部分cache沒辦法改
應該等到PR被merge就會不見了

File: config/secrets.yml (hunk 6 above)
Comment (khiav223577) [reply config/secrets.yml]: 人家都叫 secret 了

Comment (haopingfan) [reply config/secrets.yml]: 那他幹嘛不 default 在 gitignore 裡 haha

Comment (haopingfan) [reply config/secrets.yml]: 那他幹嘛不 default 在 gitignore 裡 hahahaha
更正囉~~ 刪光光了~~
謝謝神明~~

Comment (haopingfan) [reply config/secrets.yml]: 那他幹嘛不 default 在 gitignore 裡 hahahaha
喔對 有些repo是private的 可以放
更正囉~~ 刪光光了~~
謝謝神明~~

Comment (khiav223577) [reply config/secrets.yml]: private 也不能放吧 XD
預設在 gitignore 內呀～～
https://github.com/rails/rails/pull/13388

**Comment by khiav223577** [comment]
我覺得 merge 後還會在，不會不見XD

**Comment by khiav223577** [comment]
好像只能聯絡 github 客服XD 
https://help.github.com/articles/removing-sensitive-data-from-a-repository/
> Warning: Once you have pushed a commit to GitHub, you should consider any data it contains to be compromised. If you committed a password, change it! If you committed a key, generate a new one.

> This article tells you how to make commits with sensitive data unreachable from any branches or tags in your GitHub repository. However, it's important to note that those commits may still be accessible in any clones or forks of your repository, directly via their SHA-1 hashes in cached views on GitHub, and through any pull requests that reference them. You can't do anything about existing clones or forks of your repository, but you can permanently remove all of your repository's cached views and pull requests on GitHub by contacting GitHub Support.

File: config/secrets.yml (hunk 6 above)
Comment (haopingfan) [reply config/secrets.yml]: rails5.1 預設不會進gitignore了, 要跑 rails secrets:setup
https://github.com/wintermeyer/rails-5.1-book/blob/master/secrets.adoc

**Comment by haopingfan** [comment]
神明~ 我把兩串有機密資訊cache的留言刪掉囉XD
拜託不要討厭我XD

**Review by khiav223577 — APPROVED** [review]

**Review by kakas — APPROVED** [review]

**Pull request merged**